apiserver/facades/client/application/deployrepository.go

```diff
@@ -915,17 +916,25 @@ func (v *deployFromRepositoryValidator) getCharm(ctx context.Context, arg params
 		return nil, corecharm.Origin{}, nil, errors.NotSupportedf("manual deploy of the controller charm")
 	}
 
-	// Check if a charm doc already exists for this charm URL. If so, the
+	// Check if a charm already exists for this charm URL. If so, the
 	// charm has already been queued for download so this is a no-op. We
 	// still need to resolve and return back a suitable origin as charmhub
 	// may refer to the same blob using the same revision in different
 	// channels.
-	deployedCharm, err := v.state.Charm(resolvedData.URL.String())
-	if err != nil && !errors.Is(err, errors.NotFound) {
-		return nil, corecharm.Origin{}, nil, errors.Trace(err)
-	} else if err == nil {
+	deployedCharmId, err := v.applicationService.GetCharmID(ctx, applicationcharm.GetCharmArgs{
+		Name:     resolvedData.URL.Name,
+		Revision: ptr(resolvedData.URL.Revision),
```

Comment:
Could you also add the `Source` please? My PR is already landed and you have an example in https://github.com/juju/juju/blob/main/apiserver/common/charms/common.go#L71